[Concurrency] Async CC, part 1.

nate-chandler opened the pull request:
Here, the following is implemented:
- Construction of SwiftContext struct with the fields needed for calling functions.
- Allocating and deallocating these swift context via runtime calls before calling async functions and after returning from them.
- Storing arguments (including bindings and the self parameter but not including protocol fields for witness methods) and returns (both direct (however the convention is currently ignorend) and indirect).
- Wrangling errors.

Additional things that still need to be done:
- interaction between async function contexts and coroutine contexts
- yields from async functions
- partial applies of async functions

To support, a couple of structural changes were helpful:
- addition of a small EntryPointCallingConvention class hierarchy
- addition of a couple of subclasses of CallEmission

nate-chandler committed:
[Runtime] Stubbed entry points for task de/alloc.

nate-chandler committed:
[NFC] Documented IRGenFunction::IndirectReturn.

nate-chandler committed:
[Concurrency] Pass "async" into SILFunctionTypes.

This reverts commit 3a48396110d2639d7b90970bc0369a559adbc0c1.

nate-chandler commented:
@swift-ci please test

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1515,13 +1705,323 @@ void irgen::extractScalarResults(IRGenFunction &IGF, llvm::Type *bodyType,
   auto call = emitCallSite();
 
+  if (getCallee().getOrigFunctionType()->isAsync()) {
+    SILFunctionConventions fnConv(getCallee().getOrigFunctionType(),
```
I think this should be the substituted function type.

nate-chandler replied:
Fixed.

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1515,13 +1705,323 @@ void irgen::extractScalarResults(IRGenFunction &IGF, llvm::Type *bodyType,
+      Address fieldAddr =
+          fieldLayout.project(IGF, dataAddr, /*offsets*/ llvm::None);
+      auto &fieldTI = fieldLayout.getType();
+      cast<LoadableTypeInfo>(fieldTI).loadAsCopy(IGF, fieldAddr,
```
This should be loadAsTake (+0 load). Any return calling convention ownership operation should have happened as part of SIL emission i.e it is not the `return %owned_value : $SomeClass` that performs the +1. Rather it is going to be some `%owned_value = copy_value ` or some other instructions before the return.

We try to use the SILFunctionConventions abstraction rather the directly operating on SILFunctionType API. I would rather use the `SILFunctionConventions(getCallee().getSubstFunctionType(), Module).getDirectSILResults())` from above's `fnConv`  here. Sorry for not noticing/forgetting that when we looked at the code yesterday.

nate-chandler replied:
Thanks for the explanation!  Fixed.

aschwaighofer commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -1116,6 +1120,193 @@ class IRGenSILFunction :
+    bool dependsOnContextLifetime;
+    loadArgumentFromContext(explosion, layout.getLocalContextType(), ti, addr,
+                            layout.getLocalContextConvention(), IGF,
+                            IGF.CurSILFn->getLoweredFunctionType(),
```
I am confused by this line. Shouldn't that be `/*isNoEscape*true`?

nate-chandler replied:
It definitely should be.  Fixed.

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -719,6 +720,88 @@ static unsigned findSinglePartiallyAppliedParameterIndexIgnoringEmptyTypes(
+    break;
+  case ParameterConvention::Direct_Owned:
+    // Copy the value out at +1.
+    cast<LoadableTypeInfo>(fieldTI).loadAsCopy(subIGF, fieldAddr, param);
```
I think this should be a take for async context since the caller of the async function guarantees the +1.
Also I assume we never actually hit any of the indirect cases (except for indirect_inout) for async contexts. Because they all should map to a CanInout(AST.Type). Can we assert this? and prove me right or wrong.

aschwaighofer replied:
Basically the rule I have in mind is: The SwiftContext does not take ownership of any arguments or result. The ownership is held by the allocated (alloc_stack) memory location in the caller or by the reference passed by the caller. Therefore all indirect arguments should be just passing the address allocated in the caller and direct arguments should just extract the value out of the SwiftContext at +0.

nate-chandler replied:
Thanks again for the thorough explanation!  These functions are now removed and moving values into and out of the context generate no ARC traffic.

nate-chandler committed:
[Concurrency] First steps towards async CC.

Here, the following is implemented:
- Construction of SwiftContext struct with the fields needed for calling
  functions.
- Allocating and deallocating these swift context via runtime calls
  before calling async functions and after returning from them.
- Storing arguments (including bindings and the self parameter but not
  including protocol fields for witness methods) and returns (both
  direct and indirect).
- Calling async functions.

Additional things that still need to be done:
- protocol extension methods
- protocol witness methods
- storing yields
- partial applies

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please asan test

nate-chandler commented:
@swift-ci please test windows

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows

nate-chandler commented:
@swift-ci please asan test

nate-chandler commented:
@swift-ci please test

aschwaighofer reviewed (approved):
LGTM.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows

nate-chandler commented:
@swift-ci please test linux

nate-chandler commented:
@swift-ci please clean test linux

nate-chandler commented:
@swift-ci please test linux

nate-chandler commented:
@swift-ci please test linux

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows
